File: app/src/main/java/com/rangerdie/runzombiez/audio/AudioEngine.kt
package com.rangerdie.runzombiez.audio

import android.content.Context
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import com.rangerdie.runzombiez.mission.AudioChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

private const val FADE_STEP_MS = 30L

/**
 * Owns one ExoPlayer per audio channel so music, voice, sfx, and ambience can
 * layer and cross-fade independently (spec section 9/28). Built on Media3/ExoPlayer
 * with a custom [PanAudioProcessor] on the SFX channel for spatial zombie audio
 * (spec section 10).
 */
class AudioEngine(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob())
    private var fadeJobs = mutableMapOf<AudioChannel, Job>()

    private val panProcessor = PanAudioProcessor()

    private val attributes = AudioAttributes.Builder()
        .setContentType(C.AUDIO_CONTENT_TYPE_MOVIE)
        .setUsage(C.USAGE_MEDIA)
        .build()

    private val players: Map<AudioChannel, ExoPlayer> = AudioChannel.entries.associateWith { channel ->
        val builder = if (channel == AudioChannel.SFX) {
            ExoPlayer.Builder(context, PanningRenderersFactory(context, arrayOf(panProcessor)))
        } else {
            ExoPlayer.Builder(context)
        }
        builder.build().apply {
            setAudioAttributes(attributes, /* handleAudioFocus= */ channel == AudioChannel.MUSIC)
            volume = 0f
            // Missing/unresolvable audio assets (expected until docs/ASSETS.md is filled in) surface
            // here rather than crashing playback — the mission timeline keeps running silently.
            addListener(object : Player.Listener {
                override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
                    android.util.Log.w("AudioEngine", "[$channel] playback error: ${error.message}")
                }
            })
        }
    }

    /** Plays [assetPath] (relative to assets/audio/) on [channel], replacing whatever is currently playing there. */
    fun play(
        channel: AudioChannel,
        assetPath: String,
        loop: Boolean = false,
        fadeInMs: Long = 400L,
        targetVolume: Float = 1f,
        pan: Float = 0f
    ) {
        val player = players.getValue(channel)
        if (channel == AudioChannel.SFX) panProcessor.pan = pan

        player.setMediaItem(MediaItem.fromUri("asset:///audio/$assetPath"))
        player.repeatMode = if (loop) Player.REPEAT_MODE_ONE else Player.REPEAT_MODE_OFF
        player.prepare()
        player.play()
        fadeTo(channel, targetVolume, fadeInMs)
    }

    fun stop(channel: AudioChannel, fadeOutMs: Long = 400L) {
        fadeTo(channel, 0f, fadeOutMs) {
            players.getValue(channel).stop()
        }
    }

    fun stopAll(fadeOutMs: Long = 500L) {
        AudioChannel.entries.forEach { stop(it, fadeOutMs) }
    }

    /** Temporarily lowers a channel's volume (e.g. ducking music under Haven Base dialogue). */
    fun duck(channel: AudioChannel, to: Float = 0.25f, durationMs: Long = 250L) {
        fadeTo(channel, to, durationMs)
    }

    fun restore(channel: AudioChannel, to: Float = 1f, durationMs: Long = 250L) {
        fadeTo(channel, to, durationMs)
    }

    fun setPan(pan: Float) {
        panProcessor.pan = pan
    }

    private fun fadeTo(channel: AudioChannel, target: Float, durationMs: Long, onComplete: (() -> Unit)? = null) {
        fadeJobs[channel]?.cancel()
        fadeJobs[channel] = scope.launch {
            val player = players.getValue(channel)
            val start = player.volume
            if (durationMs <= 0L) {
                player.volume = target
            } else {
                val steps = (durationMs / FADE_STEP_MS).coerceAtLeast(1)
                for (i in 1..steps) {
                    player.volume = start + (target - start) * (i / steps.toFloat())
                    delay(FADE_STEP_MS)
                }
            }
            onComplete?.invoke()
        }
    }

    fun release() {
        fadeJobs.values.forEach { it.cancel() }
        players.values.forEach { it.release() }
    }
}

File: web/js/audio-engine.js
// Layered audio engine: one HTMLAudioElement + GainNode + StereoPannerNode per
// channel (MUSIC / VOICE / SFX / AMBIENCE), so tracks can play simultaneously
// and cross-fade independently. Web Audio's native StereoPannerNode gives us
// real spatial pan for zombie encounters for free — no custom DSP needed.

const CHANNELS = ["MUSIC", "VOICE", "SFX", "AMBIENCE"];
const FADE_STEP_MS = 30;

export class AudioEngine {
  constructor(basePath = "audio/") {
    this.basePath = basePath;
    this.ctx = null;
    this.nodes = {};
    this.fadeTimers = {};
  }

  // AudioContext must be created/resumed from a user gesture (browser autoplay policy).
  ensureContext() {
    if (this.ctx) return;
    this.ctx = new (window.AudioContext || window.webkitAudioContext)();
    for (const channel of CHANNELS) {
      const element = new Audio();
      element.crossOrigin = "anonymous";
      element.preload = "auto";
      const source = this.ctx.createMediaElementSource(element);
      const gain = this.ctx.createGain();
      const panner = this.ctx.createStereoPanner();
      gain.gain.value = 0;
      source.connect(gain).connect(panner).connect(this.ctx.destination);
      this.nodes[channel] = { element, gain, panner };
    }
  }

  async resume() {
    this.ensureContext();
    if (this.ctx.state === "suspended") await this.ctx.resume();
  }

  play(channel, assetPath, { loop = false, fadeInMs = 400, volume = 1, pan = 0 } = {}) {
    this.ensureContext();
    const node = this.nodes[channel];
    node.panner.pan.value = Math.max(-1, Math.min(1, pan));
    node.element.loop = loop;
    node.element.src = this.basePath + assetPath;
    node.element.currentTime = 0;
    node.element.play().catch((err) => {
      // Missing/unresolvable audio assets are expected until real audio is
      // produced (see docs/ASSETS.md) — log and keep the mission running silently.
      console.warn(`[AudioEngine] ${channel} could not play "${assetPath}":`, err.message);
    });
    this.fadeTo(channel, volume, fadeInMs);
  }

  stop(channel, fadeOutMs = 400) {
    this.fadeTo(channel, 0, fadeOutMs, () => {
      const node = this.nodes[channel];
      if (node) node.element.pause();
    });
  }

  stopAll(fadeOutMs = 500) {
    CHANNELS.forEach((c) => this.stop(c, fadeOutMs));
  }

  duck(channel, to = 0.25, durationMs = 250) {
    this.fadeTo(channel, to, durationMs);
  }

  restore(channel, to = 1, durationMs = 250) {
    this.fadeTo(channel, to, durationMs);
  }

  fadeTo(channel, target, durationMs, onComplete) {
    const node = this.nodes[channel];
    if (!node) return;
    clearInterval(this.fadeTimers[channel]);

    const start = node.gain.gain.value;
    if (durationMs <= 0) {
      node.gain.gain.value = target;
      onComplete?.();
      return;
    }
    const steps = Math.max(1, Math.round(durationMs / FADE_STEP_MS));
    let i = 0;
    this.fadeTimers[channel] = setInterval(() => {
      i += 1;
      node.gain.gain.value = start + (target - start) * (i / steps);
      if (i >= steps) {
        clearInterval(this.fadeTimers[channel]);
        onComplete?.();
      }
    }, FADE_STEP_MS);
  }

  release() {
    CHANNELS.forEach((c) => clearInterval(this.fadeTimers[c]));
    this.ctx?.close();
  }
}
